In [188]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Importing the Dataset

In [189]:
dataset.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## Finding the number of null values in dataset

In [190]:
dataset=pd.read_csv('train.csv')
dataset.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Geting the desciption of dataset

In [191]:
dataset.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Droping the Feature 'Name', 'Ticket' and 'Cabin' from dataset

In [192]:
dataset_1=dataset.drop(['Name','Ticket','Cabin'],axis=1)
dataset_1.head(10)

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S
5            6         0       3    male   NaN      0      0   8.4583        Q
6            7         0       1    male  54.0      0      0  51.8625        S
7            8         0       3    male   2.0      3      1  21.0750        S
8            9         1       3  female  27.0      0      2  11.1333        S
9           10         1       2  female  14.0      1      0  30.0708        C

## Splitting the Dataset into two Matrix X and y

In [193]:
dataset_X=dataset_1.drop('Survived',axis=1)
dataset_y=dataset_1['Survived']
X=dataset_X.iloc[:,1:].values
y=dataset_y.values

In [194]:
X[0:10,:]

array([[3, 'male', 22.0, 1, 0, 7.25, 'S'],
       [1, 'female', 38.0, 1, 0, 71.2833, 'C'],
       [3, 'female', 26.0, 0, 0, 7.925, 'S'],
       [1, 'female', 35.0, 1, 0, 53.1, 'S'],
       [3, 'male', 35.0, 0, 0, 8.05, 'S'],
       [3, 'male', nan, 0, 0, 8.4583, 'Q'],
       [1, 'male', 54.0, 0, 0, 51.8625, 'S'],
       [3, 'male', 2.0, 3, 1, 21.075, 'S'],
       [3, 'female', 27.0, 0, 2, 11.1333, 'S'],
       [2, 'female', 14.0, 1, 0, 30.0708, 'C']], dtype=object)

In [195]:
y[0:5]

array([0, 1, 1, 1, 0])

## Now as we saw we have only 714 values in Age feature, we have to impute the Age

In [196]:
from sklearn.impute import SimpleImputer
imp=SimpleImputer()
X[:,2:3]=imp.fit_transform(X[:,2:3])

In [197]:
X[0:10,:]

array([[3, 'male', 22.0, 1, 0, 7.25, 'S'],
       [1, 'female', 38.0, 1, 0, 71.2833, 'C'],
       [3, 'female', 26.0, 0, 0, 7.925, 'S'],
       [1, 'female', 35.0, 1, 0, 53.1, 'S'],
       [3, 'male', 35.0, 0, 0, 8.05, 'S'],
       [3, 'male', 29.69911764705882, 0, 0, 8.4583, 'Q'],
       [1, 'male', 54.0, 0, 0, 51.8625, 'S'],
       [3, 'male', 2.0, 3, 1, 21.075, 'S'],
       [3, 'female', 27.0, 0, 2, 11.1333, 'S'],
       [2, 'female', 14.0, 1, 0, 30.0708, 'C']], dtype=object)

## Now we will change the categorical features
### First we have to Encode the features 'Sex'(Male - 1, Female - 0)  using LabelEncoder

In [198]:
from sklearn.preprocessing import LabelEncoder
LE = LabelEncoder()
X[:,1]=LE.fit_transform(X[:,1])


In [199]:
X.shape

(891, 7)

In [200]:
X.size

6237

In [201]:
X[0:5,:]

array([[3, 1, 22.0, 1, 0, 7.25, 'S'],
       [1, 0, 38.0, 1, 0, 71.2833, 'C'],
       [3, 0, 26.0, 0, 0, 7.925, 'S'],
       [1, 0, 35.0, 1, 0, 53.1, 'S'],
       [3, 1, 35.0, 0, 0, 8.05, 'S']], dtype=object)

## Imputing the NaN valueswith most frequent value in Emarked Column

In [202]:
X[0:5,6:]

array([['S'],
       ['C'],
       ['S'],
       ['S'],
       ['S']], dtype=object)

In [203]:
imp1=SimpleImputer(strategy="most_frequent")
X[:,6:]=imp1.fit_transform(X[:,6:])

## Encoding the value in Embarked Column S-2,C-0,Q-1

In [204]:
LE2 = LabelEncoder()
X[:,6]=LE2.fit_transform(X[:,6])

In [185]:
X[0:5]

array([[3, 1, 22.0, 1, 0, 7.25, 2],
       [1, 0, 38.0, 1, 0, 71.2833, 0],
       [3, 0, 26.0, 0, 0, 7.925, 2],
       [1, 0, 35.0, 1, 0, 53.1, 2],
       [3, 1, 35.0, 0, 0, 8.05, 2]], dtype=object)

## One Hot Encoding

In [205]:
from sklearn.preprocessing import OneHotEncoder
ohe=OneHotEncoder(categorical_features=[0])
X=ohe.fit_transform(X).toarray()

/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/_encoders.py:368: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/_encoders.py:390: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [206]:
X[0:5]

array([[ 0.    ,  0.    ,  1.    ,  1.    , 22.    ,  1.    ,  0.    ,
         7.25  ,  2.    ],
       [ 1.    ,  0.    ,  0.    ,  0.    , 38.    ,  1.    ,  0.    ,
        71.2833,  0.    ],
       [ 0.    ,  0.    ,  1.    ,  0.    , 26.    ,  0.    ,  0.    ,
         7.925 ,  2.    ],
       [ 1.    ,  0.    ,  0.    ,  0.    , 35.    ,  1.    ,  0.    ,
        53.1   ,  2.    ],
       [ 0.    ,  0.    ,  1.    ,  1.    , 35.    ,  0.    ,  0.    ,
         8.05  ,  2.    ]])

In [207]:
ohe1=OneHotEncoder(categorical_features=[8])
X=ohe1.fit_transform(X).toarray()

/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/_encoders.py:368: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/_encoders.py:390: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [208]:
X[0:5]

array([[ 0.    ,  0.    ,  1.    ,  0.    ,  0.    ,  1.    ,  1.    ,
        22.    ,  1.    ,  0.    ,  7.25  ],
       [ 1.    ,  0.    ,  0.    ,  1.    ,  0.    ,  0.    ,  0.    ,
        38.    ,  1.    ,  0.    , 71.2833],
       [ 0.    ,  0.    ,  1.    ,  0.    ,  0.    ,  1.    ,  0.    ,
        26.    ,  0.    ,  0.    ,  7.925 ],
       [ 0.    ,  0.    ,  1.    ,  1.    ,  0.    ,  0.    ,  0.    ,
        35.    ,  1.    ,  0.    , 53.1   ],
       [ 0.    ,  0.    ,  1.    ,  0.    ,  0.    ,  1.    ,  1.    ,
        35.    ,  0.    ,  0.    ,  8.05  ]])

## Splitting the Dataset

In [209]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

## Feature Scaling 

In [212]:
from sklearn.preprocessing import StandardScaler
SC_X=StandardScaler()
X_train=SC_X.fit_transform(X_train)
X_test=SC_X.transform(X_test)

# IT'S Time to Prepare the Model

## Logistic Regression

In [213]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train,y_train)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=0, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [214]:
y_pred=classifier.predict(X_test)

In [215]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)

In [217]:
cm

array([[92, 18],
       [18, 51]])

This is what we got out of confusion Matrix. We have 18+18 wrong values predicted.

## Logistic when leaving one column for each encoded feature

In [221]:
X2=X[:,[1,2,4,5,6,7,8,9,10]]

In [222]:
from sklearn.model_selection import train_test_split
X_train2,X_test2,y_train2,y_test2=train_test_split(X2,y,test_size=0.2,random_state=0)

In [225]:
SC_X2=StandardScaler()
X_train2=SC_X2.fit_transform(X_train2)
X_test2=SC_X2.transform(X_test2)

In [226]:
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train2,y_train2)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=0, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [227]:
y_pred2=classifier.predict(X_test2)

In [228]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test2,y_pred2)

In [229]:
cm

array([[93, 17],
       [18, 51]])

Still its not working as expected. So, we will try another model

# SVM (Guassian)

In [230]:
from sklearn.svm import SVC
classifier_SVM = SVC(kernel = 'rbf',random_state=0)
classifier_SVM.fit(X_train2,y_train2)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=0,
  shrinking=True, tol=0.001, verbose=False)

In [231]:
y_pred_SVM=classifier.predict(X_test2)

In [232]:
cm_SVM=confusion_matrix(y_test2,y_pred_SVM)

In [233]:
cm_SVM

array([[93, 17],
       [18, 51]])